module/zfs/dmu_objset.c

```diff
@@ -965,16 +995,38 @@ dmu_objset_create_sync(void *arg, dmu_tx_t *tx)
 		    doca->doca_cred, tx);
 	}
 
+	/*
+	 * doca_userfunc() may write out some data that needs to be encrypted
+	 * if the dataset is encrypted (specifically the root directory).
+	 * However, this data won't normally be written out by the zio layer
+	 * until after this sync function has returned so we cannot remove
+	 * the key mappings until then. In this case, we call dmu_objset_sync()
+	 * to ensure the encrypted data is written out before we release the
+	 * dataset.
```

Comment:
will fix. I need to edit this comment anyway since we actually need to call `dmu_objset_sync()` twice and I should have an explanation for that.